### PR TITLE
Partial ActiveSupport::Concern support through a rewriter

### PR DESCRIPTION
**Pull request by KaanOzkan**
Blocked by: https://github.com/sorbet/sorbet/pull/3959

Introduced ActiveSupport::Concern support. The rewriter supports using `class_methods` and `module ClassMethods` to include both instance methods and class methods into a class.

Rewriter works in 2 stages. First stage is triggering. We need to trigger the rewrite whenever `extend ActiveSupport::Concern` is found on the tree. Solution utilizes the existing tree traversal in `rewriter.cc` instead of doing an additional traversal inside the rewriter for performance reasons.

Second stage of the rewriter replaces nodes similar to other rewriters. It operates on a class definition and preserves the part of the body that is out of scope for the rewriter.

### Motivation

This is another step in enabling Sorbet to provide support for Rails constructs. In Rails projects ActiveSupport::Concern is heavily used since it simplifies the code: https://api.rubyonrails.org/classes/ActiveSupport/Concern.html#method-i-class_methods

Sorbet currently does not understand methods defined inside `class_methods` and those methods are not exposed to the class that includes the module.

### Test plan

See included automated tests. I tried to cover as many cases as I can think of. They are all in the same file for easier navigation.

To be merged after https://github.com/sorbet/sorbet/pull/3424

### CODE REVIEW

File: core/tools/generate_names.cc
Comment (KaanOzkan) [inline core/tools/generate_names.cc:282]: By other PR I'm talking about https://github.com/sorbet/sorbet/pull/3424/files#diff-8b857b0dc241f308e4dfaf9cad799beeL282 . I want to name this `"classMethods"` once that is merged.

File: rewriter/rewriter.cc
Comment (KaanOzkan) [inline rewriter/rewriter.cc:131]: This check can be inserted inside the rewriter and let the rewriter traverse the tree as well. We thought this way would be a good performance improvement however, I don't know how much it actually affects performance. I wanted to point it out since this is unique to this rewriter.

File: rewriter/Concern.cc
Comment (Morriar) [inline rewriter/Concern.cc:80]:
```suggestion
// This rewriter is triggered when we confirm the ClassDef is extending `ActiveSupport::Concern`.
// Then, for a given ClassDef node it traverses the members (rhs) to identify `class_methods` calls
// or `module ClassMethods` definitions. For the former it transforms it into Ruby code that Sorbet can understand.
// It also handles cases where both `class_methods` and `module ClassMethods` are used.
// Knowledge of `class_methods` and its source code will help you understand the steps taken here.
// See https://api.rubyonrails.org/classes/ActiveSupport/Concern.html#method-i-class_methods.
```

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (jez) [inline test/testdata/rewriter/concern.rb.rewrite-tree.exp:16]: This references a constant before it's defined. Can we move this to after the definition of `module ClassMethods`?

Comment (KaanOzkan) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp]: Done. 

Ordering of the nodes is not preserved perfectly compared to the input due to the implementation itself. I didn't think it was necessary and I think it would result in a more complicated logic to achieve it. I did the minimal change to reference the constant after its defined.

File: rewriter/rewriter.cc
Comment (jez) [reply rewriter/rewriter.cc:140]: Yeah I'm not too worried about this for performance. Would you mind moving this to be the first thing that `Concern::run` does? (Given that it's already operating on the whole ClassDef it shouldn't be too much work).

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (jez) [inline test/testdata/rewriter/concern.rb.rewrite-tree.exp:20]: This is going to silence some exceptions that happen at runtime.

```ruby
module Foo
  extend ActiveSupport::Concern
  class_methods do
    def a_class_method
    end
  end

  abstract! # exception at runtime, no static error
end
```

I'm not sure of a way to handle that.

Can you add this as a test, and confirm that this is a tradeoff that you're ok with? Stripe's codebase will not be using this so I'm fine with this as long as you are.

File: test/testdata/rewriter/concern.rb
Comment (jez) [inline test/testdata/rewriter/concern.rb:130]: Can you add a test for what happens if `class_methods` appears twice in the body?

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:143]:
```suggestion
    if (firstArg == nullptr) {
        return false;
    }
    auto *scope = ast::cast_tree<ast::UnresolvedConstantLit>(firstArg->scope);
    if (scope == nullptr) {
```

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:148]: What happens if I write this:

```ruby
extend DefinitelyNot::ActiveSupport::Concern
```

Does this rewriter pass still fire?

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:10]:
```suggestion

namespace {

```

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:65]:
```suggestion

} // namespace

```

File: rewriter/Concern.cc
Comment (jez) [reply rewriter/Concern.cc:10]: These functions all have public (non-static) linkage, which can cause confusing linker errors if someone else happens to write a function with the same name.

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:27]: I admire your desire to factor code into small pieces, but I actually find this to be harder to read, because these functions are only called once (it looks like maybe one of these helpers is called twice). It means that instead of reading the code top-to-bottom, seeing which ast nodes are getting created and where, I have to jump all over this file. Also we're mutating a vector in the process, and that's hidden behind a function call.

Do you mind either:

- inlining all of these short functions into the place where they're called or,
- keeping them but making them return the thing they create, and doing the `stats.emplace_back(...)` calls inside the main method?

**Review by jez — CHANGES_REQUESTED** [review]
Thanks, this change looks mostly good! I have a couple comments on the implementation itself, but otherwise I'm happy to merge this.

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (KaanOzkan) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp:20]: @jez Thank you so much for bringing this up, I missed it. We don't want to make this tradeoff and happy to implement a solution as part of this PR.

We had internal discussions and we thought of adding a `T::Helpers::WithoutRuntime.mixes_in_class_methods` (similar to `T::Sig::WithoutRuntime.sig`) which has no runtime implementation (since `ActiveSupport::Concern` does its own implementation at runtime anyway).

This way instead of inserting a `extend(T::Helpers)` we can explicitly call `T::Helpers::WithoutRuntime.mixes_in_class_methods`. Let me know what do you think 😊

Comment (jez) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp:20]: Sounds good. I'm leaning towards something like `<Magic>.<mixes_in_class_methods>` because I don't know whether we want to implement this at runtime, and using `<Magic>` would make it impossible to be used outside of Sorbet internals.

If we do want to implement this at runtime, it would have to accept `self` as an argument, and then we may as well make it a class method on `T::Helpers` because it works at runtime.

Regardless, Stripe is having a global remote offsite this week, so I'll have to get back to you in a few days after my team has a chance to meet and discuss this.

File: rewriter/Concern.cc
Comment (KaanOzkan) [reply rewriter/Concern.cc]: It did fire, thank you. I opted for a similar approach as before to fix this. (`scope` is now renamed to `firstArgScope`)

```
auto *outerScope = ast::cast_tree<ast::UnresolvedConstantLit>(firstArgScope->scope);
if (outerScope != nullptr) {
    return;
}
```

File: rewriter/rewriter.cc
Comment (KaanOzkan) [reply rewriter/rewriter.cc]: This is now triggered within `Concern::run`. I repeated the usage of `typecase()` switch statement similar to `rewriter.cc`. If there is a simpler approach you recommend I can change it.

**Review by jez — APPROVED** [review]
@KaanOzkan This change looks fine to me.

I'm fine either

- landing as is, and filing a ticket or following up with a second PR to do the `<Magic>.<mixes_in_class_methods>` stuff, or
- blocking until that's implemented in this PR.

If you'd like to merge this now, let me know.

In any case, I'd like to implement and review the `<Magic>.<mixes_in_class_methods>` change as a separate PR.

**Comment by KaanOzkan** [comment]
Thanks @jez. I'd like for it to be blocked until `<Magic>.<mixes_in_class_methods>` is implemented. I don't want to silence runtime exceptions even temporarily. I'll get on that change shortly after the new year, thanks again.

**Review by jez — CHANGES_REQUESTED** [review]
https://github.com/sorbet/sorbet/pull/3468#issuecomment-749360341

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (KaanOzkan) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp:20]: @jez I started implementing this and I'm at a crossroads while trying to satisfy typechecking. I initially opted for adding a simple `Magic` class that declares `mixes_in_class_methods`in the rewriter stage. However, I'm having trouble generating the method on the **singleton class** of `Magic`.

I'm curious if that's the route you wanted to take or should I instead follow a similar approach to https://github.com/sorbet/sorbet/pull/2194/files for typechecking errors related to `Magic#mixes_in_class_methods`.

File: resolver/resolver.cc
Comment (KaanOzkan) [inline resolver/resolver.cc:772]: I couldn't find a way to check for the constant name `Magic` without the double casting, as a result it got messy here. Open to suggestions.

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (KaanOzkan) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp]: Ended up defining the method in `GlobalState.cc` similar to other `<Magic>` usages and it's simpler now.

File: resolver/resolver.cc
Comment (jez) [reply resolver/resolver.cc:772]: See my suggested changes.

Feel free to play around with them, in case you want to further simplify anything.

File: rewriter/Concern.cc
Comment (jez) [inline rewriter/Concern.cc:122]:
```suggestion
        auto magic = ast::MK::Constant(klass->loc, ast::MK::EmptyTree(), core::Symbols::Magic());
```

File: resolver/resolver.cc
Comment (jez) [inline resolver/resolver.cc:772]:
```suggestion
            if (recvAsConstantLit != nullptr && recvAsConstantLit->symbol == core::Symbols::Magic() && ) {
                auto item = ClassMethodsResolutionItem{ctx.file, ctx.owner, &send};
                this->todoClassMethods_.emplace_back(move(item));
            }
```

File: rewriter/Concern.h
Comment (jez) [inline rewriter/Concern.h:25]:
```suggestion
 *      <Magic>.mixes_in_class_methods(ClassMethods)
```

File: core/GlobalState.cc
Comment (jez) [inline core/GlobalState.cc:668]:
```suggestion
    // Synthesize <Magic>.mixes_in_class_methods(arg: T.untyped) => Void
```

**Review by jez — APPROVED** [review]
Looks good to me. Left a few suggestions. Once those are fixed we can merge.

Thanks!

**Comment by jez** [comment]
Looks like you've got a test exp file that fails, and also I think you're conflicting with #3424 (your PR that you said you wanted to be merged first)

**Comment by KaanOzkan** [comment]
@jez I have fixed the conflicts and updated the test `exp` due to latest changes.

File: test/testdata/rewriter/concern.rb.rewrite-tree.exp
Comment (jez) [reply test/testdata/rewriter/concern.rb.rewrite-tree.exp]: > Sounds good. I'm leaning towards something like `<Magic>.<mixes_in_class_methods>` because I don't know whether we want to implement this at runtime, and using `<Magic>` would make it impossible to be used outside of Sorbet internals.
> 
> If we do want to implement this at runtime, it would have to accept `self` as an argument, and then we may as well make it a class method on `T::Helpers` because it works at runtime.

@KaanOzkan Could you make the new Magic mixes_in_class_methods method take `self` as an argument?

**Comment by jez** [comment]
> In any case, I'd like to implement and review the `<Magic>.<mixes_in_class_methods>` change as a separate PR.

@KaanOzkan I was hoping to see the `mixes_in_class_methods` change as a separate PR, so that we could more easily isolate it if it ends up causing problems somehow. Would you mind splitting this into two PRs?

**Comment by KaanOzkan** [comment]
@jez Sorry I missed that. I will move the changes in `GlobalState.cc` and `SymbolRef.h` into a separate PR and clean up the commit history for this one.

That new PR can be a blocker for this one.

**Comment by KaanOzkan** [comment]
@jez Rebased on the latest changes and cleaned up the commit history

**Pull request merged**
